### PR TITLE
Add docs for FromJSONKey

### PR DESCRIPTION
**Pull request by andrewthad**

### CODE REVIEW

File: Data/Aeson/Types/FromJSON.hs [hunk 1]
```diff
@@ -321,33 +321,53 @@ class FromJSON a where
...
+--   For performance reasons, these exist as three options instead of one.
 data FromJSONKeyFunction a
-    = FromJSONKeyCoerce (CoerceText a)
-    | FromJSONKeyText (Text -> a)
-    | FromJSONKeyTextParser (Text -> Parser a)
-    | FromJSONKeyValue (Value -> Parser a)
+    = FromJSONKeyCoerce (CoerceText a) -- ^ uses 'coerce' ('unsafeCoerce' in older GHCs)
+    | FromJSONKeyText (Text -> a) -- ^ conversion from 'Text' that always succeeds
+    | FromJSONKeyTextParser (Text -> Parser a) -- ^ conversion from 'Text' that may fail
+    | FromJSONKeyValue (Value -> Parser a) -- ^ conversion for non-textual keys
```
Comment (phadej) [inline Data/Aeson/Types/FromJSON.hs]: I'd add haddocks under the options:

```hs
    = FromJSONKeyCoerce (CoerceText a)
      -- ^ uses 'coerce' ('unsafeCoerce' in older GHCs)
    | FromJSONKeyText (Text -> a)
      -- ^ conversion from 'Text' that always succeeds
    | FromJSONKeyTextParser (Text -> Parser a)
      -- ^ conversion from 'Text' that may fail
    | FromJSONKeyValue (Value -> Parser a)
      -- ^ conversion for non-textual keys
 ```

Comment (andrewthad) [inline Data/Aeson/Types/FromJSON.hs]: Good call. I've changed this in the newest commit.

**Comment by bergmark** [comment]
LGTM, but needs a rebase

**Comment by andrewthad** [comment]
rebased

**Comment by bergmark** [comment]
awesome

**Pull request merged**